flink-runtime/src/main/java/org/apache/flink/runtime/rest/handler/AbstractRestHandler.java

```diff
@@ -106,6 +113,21 @@ protected AbstractRestHandler(
 		}
 	}
 
+	private CompletableFuture<P> handleRequestInternal(final HandlerRequest<R, M> handlerRequest, final T gateway) {
+		inFlightRequestTracker.registerRequest();
+
+		CompletableFuture<P> response;
+		try {
+			response = handleRequest(handlerRequest, gateway);
+		} catch (RestHandlerException e) {
+			response = FutureUtils.completedExceptionally(e);
+		} catch (Throwable t) {
+			inFlightRequestTracker.deregisterRequest();
+			throw t;
```

Comment:
you could fail the future here, then we wouldn't call `deregisterRequest` from 2 places

Reply:
Logic extracted to superclass.